Scripts/Weapon/DashSkill.cs

```diff
@@ -0,0 +1,245 @@
+            foreach (PhysicsBody body in sideBodies)
+            {
+                if (!(body is Enemy))
+                    continue;
+
+                if (!(body as Enemy).HasStatus<MarkStatus>())
+                    continue;
+                Vector3 bodyDisplacement = body.GlobalTransform.origin - GlobalTransform.origin;
+                Plane upFacingPlane = new Plane(GlobalTransform.basis.y, 0);
+                Vector3 flattenedBodyDisplacement = upFacingPlane.Project(bodyDisplacement);
+
+                float degreesDifference = Mathf.Rad2Deg(forward.AngleTo(flattenedBodyDisplacement));
+                if (closestEnemy == null || degreesDifference < lowestDegreesDifference)
+                {
+                    closestEnemy = body as Enemy;
+                    lowestDegreesDifference = degreesDifference;
+                }
+            }
+        }
+
+        return closestEnemy;
+    }
+
+    protected override void OnAttackButtonPressed()
+    {
+        showTargetHint = true;
+    }
+
+    protected override void OnAttackButtonReleased()
+    {
+        showTargetHint = false;
+        if (isRunning)
+            return;
+        Vector3 currentLocation = GlobalTransform.origin;
+        Vector3 targetLocation;
+        if (targetEnemy == null)
+        {
+            targetLocation = currentLocation + range * (-GlobalTransform.basis.z);
+        }
+        else
+        {
+            Vector3 enemyLocation = targetEnemy.GlobalTransform.origin;
+            Dictionary ray = GetWorld().DirectSpaceState.IntersectRay(currentLocation, enemyLocation, collisionMask: 4);
+            Vector3 enemyBoundary = (Vector3)ray["position"];
+
+            // Go right up beside enemy
+            targetLocation = enemyBoundary + player.Scale * GlobalTransform.basis.z;
+        }
+
+        Transform currentTransform = player.GlobalTransform;
+        Transform targetTransform = new Transform(player.GlobalTransform.basis, targetLocation);
+
+        float duration = (targetLocation - currentTransform.origin).Length() / dashSpeed;
+
+        Tween tween = new Tween();
+        AddChild(tween);
+        tween.InterpolateProperty(player, "GlobalTransform", currentTransform, targetTransform, duration, transType: Tween.TransitionType.Quart);
+        tween.InterpolateCallback(this, duration, "OnFinished");
+        tween.Start();
+
+        player.IsMovementLocked = true;
+        isRunning = true;
+    }
+
+    public void OnFinished()
+    {
+        if (targetEnemy != null)
+        {
+            Vector3 targetEnemyDirection = (targetEnemy.GlobalTransform.origin - GlobalTransform.origin).Normalized();
+            Vector3 accelerationOntargetEnemy = 50.0f * targetEnemyDirection + 40.0f * Vector3.Up;
+            targetEnemy.Velocity += accelerationOntargetEnemy;
+            targetEnemy.Damage(damage);
+
+            foreach (PhysicsBody body in impactArea.GetOverlappingBodies())

```

Comment:
Is there a chance that target enemy is part of this loop? Will it get damaged twice?

Reply:
very good point